decryptor/mysql/prepared_statement_sql_observer.go

```diff
@@ -0,0 +1,519 @@
+
+	var preparedStatementName = executeQuery.PreparedStatementName.String()
+
+	_, err := e.proxyHandler.registry.StatementByID(preparedStatementName)
+	if err != nil {
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			WithError(err).Errorln("PreparedStatement not present in registry")
+		return nil, false, ErrStatementNotPresentInRegistry
+	}
+
+	querySetting, ok := e.proxyHandler.registry.QuerySettingsByID(preparedStatementName)
+	if ok {
+		logrus.Debugln("Set query settings from registry")
+		e.querySelectSettings = querySetting
+	}
+
+	return base.NewOnQueryObjectFromStatement(executeQuery, e.parser), true, nil
+}
+
+func (e *PreparedStatementsQuery) onDeallocate(ctx context.Context, deallocateQuery *sqlparser.DeallocatePrepare) (base.OnQueryObject, bool, error) {
+	logrus.Debugln("PreparedStatementsQuery.Deallocate")
+
+	var preparedStatementName = deallocateQuery.PreparedStatementName.String()
+
+	if _, err := e.proxyHandler.registry.StatementByID(deallocateQuery.PreparedStatementName.String()); err != nil {
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			WithError(err).Errorln("PreparedStatement not present in registry")
+		return nil, false, ErrStatementNotPresentInRegistry
+	}
+
+	e.proxyHandler.registry.DeleteStatementByID(preparedStatementName)
+	return nil, false, nil
+}
+
+func (e *PreparedStatementsQuery) onQuery(ctx context.Context, preparedStatementName string, stmt sqlparser.Statement) (base.OnQueryObject, bool, error) {
+	var query = sqlparser.String(stmt)
+	var preparedStatement = NewPreparedStatement(0, 0, query, stmt)
+
+	preparedStatement.name = preparedStatementName
+	e.proxyHandler.registry.AddStatement(preparedStatement)
+
+	queryObj := base.NewOnQueryObjectFromQuery(query, e.parser)
+	changedObject, changed, err := e.proxyHandler.queryObserverManager.OnQuery(ctx, queryObj)
+	if err != nil {
+		logrus.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorEncryptQueryData).Errorln("Error occurred in OnQuery handler")
+		return nil, false, err
+	}
+
+	switch query := stmt.(type) {
+	case *sqlparser.Select:
+		querySetting, err := e.parseQuerySettings(ctx, query)
+		if err != nil {
+			logrus.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorEncryptQueryData).Errorln("Failed to parse querySettings for select in Prepare")
+			return nil, false, err
+		}
+		e.proxyHandler.registry.AddQuerySettings(preparedStatementName, querySetting)
+	}
+
+	return changedObject, changed, nil
+}
+
+// OnBind just a stub method of QueryObserver of  PreparedStatementsQuery
+func (e *PreparedStatementsQuery) OnBind(ctx context.Context, statement sqlparser.Statement, values []base.BoundValue) ([]base.BoundValue, bool, error) {
+	logrus.Debugln("PreparedStatementsQuery.OnBind")
+	return values, false, nil
+}
+
+func filterTableExpressions(statement sqlparser.Statement) (sqlparser.TableExprs, error) {
```

Comment:
why not use this function from acra/encryptor/searchable_query_filter.go? or move from there to the decryptor/base package to re-use in 2 packages?